### PR TITLE
adapter: emergency fix for misordered sinks

### PR DESCRIPTION
**Pull request by benesch**
ALTER SINK can result in sinks that depend on objects with larger IDs, which violates a invariant that is checked (implicitly and explicitly) in several places throughout the catalog.

The fixes in this patch are surgical hacks to disarm the panic in production. We'll need to follow this up with something cleaner, but doing so is blocked on a broader philosophical discussion about whether the ID order invariant is something we want to keep.

### Motivation

  * This PR fixes a recognized bug: #incident-131.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by def-** [comment]
New panic backtrace:
```
thread 'coordinator' panicked at src/adapter/src/catalog/state.rs:1116:64:
no entry found for key
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_display
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:263:5
   3: core::option::expect_failed
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/option.rs:1994:5
   4: <core::option::Option<&mz_catalog::memory::objects::Schema>>::expect
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/option.rs:895:21
   5: <alloc::collections::btree::map::BTreeMap<mz_sql::names::SchemaId, mz_catalog::memory::objects::Schema> as core::ops::index::Index<&mz_sql::names::SchemaId>>::index
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/alloc/src/collections/btree/map.rs:2333:23
   6: <mz_adapter::catalog::state::CatalogState>::get_schema
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/state.rs:1116:64
   7: <mz_adapter::catalog::state::CatalogState>::pack_item_update
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/builtin_table_updates.rs:451:26
   8: <mz_adapter::catalog::state::CatalogState>::generate_builtin_table_update
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/apply.rs:1003:44
   9: <mz_adapter::catalog::state::CatalogState>::apply_updates_inner
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/apply.rs:176:33
  10: <mz_adapter::catalog::state::CatalogState>::apply_updates
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/apply.rs:141:17
  11: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/transact.rs:511:44
  12: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  13: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/transact.rs:443:5
  14: <mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/transact.rs:404:10
  15: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  16: <mz_adapter::catalog::Catalog>::transact::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/catalog/transact.rs:340:5
  17: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:517:14
  18: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  19: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:182:5
  20: <mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:120:71
  21: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  22: <mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:114:5
  23: <mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:77:14
  24: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  25: <mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/ddl.rs:70:5
  26: <mz_adapter::coord::Coordinator>::sequence_drop_objects::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/sequencer/inner.rs:1186:51
  27: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::sequence_drop_objects::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  28: <mz_adapter::coord::Coordinator>::sequence_drop_objects::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/sequencer/inner.rs:1152:5
  29: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/sequencer.rs:232:86
  30: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  31: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  32: <mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/command_handler.rs:947:69
  33: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  34: <mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/command_handler.rs:503:5
  35: <mz_adapter::coord::Coordinator>::handle_execute::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/command_handler.rs:500:54
  36: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_execute::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  37: <mz_adapter::coord::Coordinator>::handle_execute::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/command_handler.rs:401:5
  38: <mz_adapter::coord::Coordinator>::handle_command::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/command_handler.rs:120:26
  39: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_command::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  40: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  41: <mz_adapter::coord::Coordinator>::message_command::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/message_handler.rs:388:34
  42: <mz_adapter::coord::Coordinator>::message_command::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/message_handler.rs:386:5
  43: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::message_command::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  44: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord/message_handler.rs:70:64
  45: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_message::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  46: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  47: <mz_adapter::coord::Coordinator>::serve::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord.rs:3080:48
  48: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  49: <tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  50: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/park.rs:281:63
  51: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:107:5
  52: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:73:5
  53: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/park.rs:281:31
  54: <tokio::runtime::context::blocking::BlockingRegionGuard>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/blocking.rs:66:9
  55: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/handle.rs:310:22
  56: tokio::runtime::context::runtime::enter_runtime::<<tokio::runtime::handle::Handle>::block_on<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}, ()>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/runtime.rs:65:16
  57: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/handle.rs:309:9
  58: mz_adapter::coord::serve::{closure#0}::{closure#2}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-03df0bb45fd3a9f85-1/materialize/test/src/adapter/src/coord.rs:3717:21
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by def-** [comment]
I added three kinds of tests to reproduce the panic.

File: src/adapter/src/catalog/apply.rs [hunk 1]
```diff
@@ -149,14 +149,23 @@ impl CatalogState {
...
         soft_assert_no_log!(
             updates.iter().map(|update| update.ts).all_equal(),
             "all timestamps should be equal: {updates:?}"
         );
 
+        // HACK: due to `ALTER SINK`, sinks can appear before the objects they
+        // depend upon. Fortunately, because sinks can never have dependencies
+        // and can never depend upon one another, to fix the topological sort,
+        // we can just always move sinks to the end.
+        updates.sort_by_cached_key(|u| match &u.kind {
```
Comment (jkosh44) [inline src/adapter/src/catalog/apply.rs:164]: I'm not familiar with this method, but wouldn't this destroy the relative sorting of everything else in `updates`?

Comment (jkosh44) [reply src/adapter/src/catalog/apply.rs:164]: Something like the following diff (which updates `sort_item_updates()`), would make me much more comfortable.

```
diff --git a/src/adapter/src/catalog/apply.rs b/src/adapter/src/catalog/apply.rs
index f8ec696394..decc626112 100644
--- a/src/adapter/src/catalog/apply.rs
+++ b/src/adapter/src/catalog/apply.rs
@@ -1571,7 +1571,13 @@ fn sort_updates_inner(updates: Vec<StateUpdate>) -> Vec<StateUpdate> {
     ) -> VecDeque<(mz_catalog::durable::Item, Timestamp, StateDiff)> {
         item_updates
             .into_iter()
-            .sorted_by_key(|(item, _ts, _diff)| item.id)
+            .sorted_by_key(|(item, _ts, _diff)| {
+                if item.create_sql.starts_with("CREATE SINK") {
+                    GlobalId::User(u64::MAX)
+                } else {
+                    item.id
+                }
+            })
             .collect()
     }
     let item_retractions = sort_item_updates(item_retractions);
```

Comment (jkosh44) [reply src/adapter/src/catalog/apply.rs]: I pushed this, we'll see how it does in CI.

**Comment by jkosh44** [comment]
@def- we're passing more tests but still failing some. Are they all new ones added in this PR? Or maybe known flakes?

**Comment by def-** [comment]
The platform-check was wrong, fixed it in a new commit.
Altering sinks seem to cause new stuck queries in parallel-workload: https://buildkite.com/materialize/nightly/builds/8920#01911344-393b-4b18-9ad7-04452d1186e8 but we can also disable altering sinks in parallel-workload for now and look at that separately.
[Testdrive 1](https://buildkite.com/materialize/test/builds/87382#01911343-fb41-46b1-9c8d-5eb94bcc171f) still seems to be missing the value after altering the sink.

File: test/testdrive/alter-sink.td [hunk 2]
```diff
@@ -107,3 +107,25 @@ SELECT pg_cancel_backend(CAST(${backend-pid} AS int4));
...
+
+# The sink will start sinking updates from `post_alter` at the timestamp that
+# the previous dataflow happens to stop. This happens pretty quickly but we
+# wait a few seconds more for good measure to avoid flaking.
+$ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration=4s
+
+> INSERT INTO post_alter VALUES ('hundred', 99);
+
+$ kafka-verify-data format=json sink=materialize.public.sink key=false
+{"before": null, "after": {"created_post_name": "hundred", "created_post_value": 99}}
```
Comment (jkosh44) [inline test/testdrive/alter-sink.td:131]: I applied this to `main` and this line fails on `main`. I would vote to just move this into a followup issue and focus on the corrupt catalog for now.

**Comment by jkosh44** [comment]
@def, I haven't read through all the added tests, but IMO we should separate out "ALTER SINK works" and "ALTER SINK corrupts the catalog". We want to fix the corrupt catalog stuff ASAP and it'll take me a bit to ramp up on ALTER SINK to figure out why it's not behaving as expected.

File: test/testdrive/alter-sink.td (hunk 2 above)
Comment (jkosh44) [reply test/testdrive/alter-sink.td:131]: Hmm, is the `before` wrong? Should it be the previous entry?

File: test/testdrive/alter-sink.td [hunk 3]
```diff
@@ -107,3 +107,25 @@ SELECT pg_cancel_backend(CAST(${backend-pid} AS int4));
 
 ! ALTER SINK wedged_sink SET FROM post_alter;
 contains:canceling statement due to user request
+
+
+# There is a meaninful difference in having an object created after the sink
```
Comment (jkosh44) [inline test/testdrive/alter-sink.td:112]:
```suggestion
# There is a meaningful difference in having an object created after the sink
```

File: test/testdrive/alter-sink.td (hunk 2 above)
Comment (jkosh44) [reply test/testdrive/alter-sink.td]: I commented this line out, @benesch or anyone that knows anything about `ALTER SINK`, are you OK merging with that commented out?

**Comment by jkosh44** [comment]
It looks like Nightlies are still failing, but it's not at all clear to me what about them are failing.

**Comment by def-** [comment]
The queries seem to get stuck when you run ALTER SINK with other DDLs in parallel. I'll disable the ALTER SINK in parallel-workload for now to get this green.

**Comment by jkosh44** [comment]
> The queries seem to get stuck when you run ALTER SINK with other DDLs in parallel. I'll disable the ALTER SINK in parallel-workload for now to get this green.

Just for future reference, how are you able to tell that from the test output?

**Comment by def-** [comment]
```
worker_0 still running: DROP SOURCE postgres_source3;
worker_1 still running: ALTER SINK "db-pw-1722618900-0"."s-5"."sink-10" SET FROM "db-pw-1722618900-0"."s-0".kafka_table1;
worker_2 still running: CREATE TABLE "db-pw-1722618900-0"."s-4"."t-9"("c-0-oid" oid NOT NULL,
    "c-1-oid" oid NOT NULL,
    "c-2-date" date DEFAULT DATE '95834-5-24' NOT NULL,
    "c-3-text" text DEFAULT 'Wy0d2JC0XotLdJe0b45mlOl6uqakRjnTFuhszL1jWwZ6N4YpvUyRYHjhQiBK5g0wAi9fuaHid6XRpjlBKRdi5DUmLlX0u6jsN0PA' NOT NULL,
    "c-4-date" date,
    "c-5-smallint" smallint,
    "c-6-bigint" bigint DEFAULT 7739634923868460187,
    "c-7-boolean" boolean NOT NULL,
    "c-8-oid" oid DEFAULT -1520279234);
worker_3 still running: POST http://localhost:6976/api/webhook/db-pw-1722618900-0/s-3/wh-5 Headers: x-mz-api-key: b'"V8n0IuOhXamnD41z2nxdeiNZZyM8nq74itBLmMPgg3tUYvhwU7pfBM9qmPvUKl7nKYjjfTr4Sy8ciBNF3yGwf9zLU5WRnFkFEFGx"', x-event-type: b'"QVBFCiJEdU41L2bkHS6iyHiXdEbJ6YQNMFo2ez6mRI6Gg8uBEUSgPCsWpt7SSbVlu3tYuOtjWECPTQWPBVIJrXWEVkd0kd8Cls60"', signature: b'"t4WKWmFvi6g27RxEth3uFGuxjJTCBKfLNUv7N9Mjrj10EN2rDzHElArAKQzxuq9PpP3MKSoYVLZPj3DxCgwVCUbwK3uyFSf0xEey"', timestamp: 2024-08-02 17:18:54.516039, ih0: b'"hZwDrFpGDnR9HHLoWK02dMhKprFG1Lu14JSENfajGXiQbmiyIngUoCbr26COME8DHfW5n6JK0rqwENOWa08kokzFrUidVseawRqS"' Body: b'9iPcgT2AzEQXNKqMySDtsQgBx6mu54xHzgi8BLcqj1jNmgUGp9YWFzWpZ4hR3vSFvyqftZSuKSpleOvTknr7N2szOc6UHkqdzGiZ'
worker_4 still running: rollback
worker_5 still running: COPY (SELECT * FROM "db-pw-1722618900-0"."s-3".table6) TO 's3://copytos3/130' WITH (AWS CONNECTION = aws_conn, FORMAT = 'csv');
worker_6 still running: commit;
worker_7 still running: CREATE SOURCE "db-pw-1722618900-1"."s-7".kafka_table14
                    IN CLUSTER "cluster-4"
                    FROM KAFKA CONNECTION materialize.public.kafka_conn (TOPIC 'data-ingest-14')
                    FORMAT AVRO
                    USING CONFLUENT SCHEMA REGISTRY CONNECTION materialize.public.csr_conn
                    ENVELOPE UPSERT
Threads have not stopped within 5 minutes, exiting hard
```

**Review by def- — APPROVED** [review]
Feel free to submit, I opened https://github.com/MaterializeInc/materialize/issues/28726 to follow up. (But it doesn't really seem to be a test issue to me, `ALTER SINK` seems to reliably cause stuck DDLs, but that's probably not introduced by this PR?)

**Comment by jkosh44** [comment]
> Feel free to submit, I opened #28726 to follow up. (But it doesn't really seem to be a test issue to me, `ALTER SINK` seems to reliably cause stuck DDLs, but that's probably not introduced by this PR?)

I doubt it, but it's probably worth checking on `main` too.

File: src/adapter/src/catalog/apply.rs (hunk 1 above)
Comment (benesch) [reply src/adapter/src/catalog/apply.rs]: > wouldn't this destroy the relative sorting of everything else in updates?

The sort method is stable so in theory it should leave everything in the same order. Your diff looks great though. Somehow I didn't notice the `sort_updates` function last night. 🙈

**Comment by benesch** [comment]
LGTM, Joe. Thanks for all the help.

**Pull request merged**